Firmware/MotorControl/pwm_input.cpp

```diff
@@ -0,0 +1,423 @@
+int tim_2_5_channel_num_to_gpio_num(int channel) {
+    #if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+        if (channel >= 1 && channel <= 4) {
+            // the channel numbers just happen to coincide with
+            // the GPIO numbers
+            return channel;
+        } else {
+            return -1;
+        }
+    #else
+        // Only ch4 is available on v3.2
+        if (channel == 4) {
+            return 4;
+        } else {
+            return -1;
+        }
+    #endif
+}
+
+// @brief Returns the TIM2 or TIM5 channel number
+// for a given GPIO number.
+uint32_t gpio_num_to_tim_2_5_channel(int gpio_num) {
+    #if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+        switch (gpio_num) {
+            case 1: return TIM_CHANNEL_1;
+            case 2: return TIM_CHANNEL_2;
+            case 3: return TIM_CHANNEL_3;
+            case 4: return TIM_CHANNEL_4;
+            default: return 0;
+        }
+    #else
+        // Only ch4 is available on v3.2
+        if (gpio_num == 4) {
+            return TIM_CHANNEL_4;
+        } else {
+            return 0;
+        }
+    #endif
+}
+
+void pwm_in_init() {
+    GPIO_InitTypeDef GPIO_InitStruct;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_PULLDOWN;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    GPIO_InitStruct.Alternate = GPIO_AF2_TIM5;
+
+    TIM_IC_InitTypeDef sConfigIC;
+    sConfigIC.ICPolarity = TIM_INPUTCHANNELPOLARITY_BOTHEDGE;
+    sConfigIC.ICSelection = TIM_ICSELECTION_DIRECTTI;
+    sConfigIC.ICPrescaler = TIM_ICPSC_DIV1;
+    sConfigIC.ICFilter = 15;
+
+#if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+    for (int gpio_num = 1; gpio_num <= PWM_IN_COUNT; ++gpio_num) {
+#else
+    int gpio_num = 4; {
+#endif  
+        // Activate timer interrupt if the pin is being used as a PWM input
+        bool pwm_dir_pin = false;
+        for (uint8_t i = 0; i < PWM_IN_COUNT; ++i) {
+            if (board_config.pwm_mappings[i].gpio_direction_pin == gpio_num) {
+                pwm_dir_pin = true;
+            }
+        }
+        if (is_endpoint_ref_valid(board_config.pwm_mappings[gpio_num - 1].endpoint) || pwm_dir_pin) {
+            GPIO_InitStruct.Pin = get_gpio_pin_by_pin(gpio_num);
+            HAL_GPIO_DeInit(get_gpio_port_by_pin(gpio_num), get_gpio_pin_by_pin(gpio_num));
+            HAL_GPIO_Init(get_gpio_port_by_pin(gpio_num), &GPIO_InitStruct);
+            HAL_TIM_IC_ConfigChannel(&htim5, &sConfigIC, gpio_num_to_tim_2_5_channel(gpio_num));
+            HAL_TIM_IC_Start_IT(&htim5, gpio_num_to_tim_2_5_channel(gpio_num));
+        }
+    }
+
+    // Create and start a periodic RTOS timer for pwm_no_pulse_check_cb()
+    osTimerDef_t pwm_check_callback_struct;
+    pwm_check_callback_struct.ptimer = &pwm_no_pulse_check_cb; // Callback function pointer
+    osTimerId pwm_check_timer = osTimerCreate(&pwm_check_callback_struct, osTimerPeriodic, NULL);
+    osTimerStart (pwm_check_timer, 5); // Period in milliseconds
```

Comment:
You're right, having the user set a nominal PWM frequency instead of trying to detect it would get rid of a lot of this logic.

The timeout for jumping from >0% duty cycle to 0% duty cycle is actually PWM_INACTIVE_TIME, set to 45ms, so it's worse than you thought.  45ms works for RC PWM, but for duty-cycle PWM it should be set relative to the nominal frequency like you said.